File: 242-valid-anagram/valid-anagram.cpp
class Solution {
public:
    bool isAnagram(string s, string t) {
        if(s.size() != t.size()) return false;

        // sort(s.begin(),s.end()); // O(nlogn)
        // sort(t.begin(), t.end());

        // for(int i=0; i<s.size(); i++){       // O(n)
        //     if(s[i] != t[i]) return false;
        // }
        // return true;

        vector<int> freq(26, 0); // frequency of letters a-z

        for(int i=0; i<s.size(); i++){
            freq[s[i] - 'a']++;
            freq[t[i] - 'a']--;
        }

        for(int count: freq){
            if(count != 0) return false;
        }

        return true;
    }
};

File: 242-valid-anagram/valid-anagram.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if(len(s) != len(t)):
            return False
        
        # counts = [0]*26
        
        # for i in range(len(s)):
        #     counts[ord(s[i]) - ord('a')] += 1
        #     counts[ord(t[i]) - ord('a')] -= 1
        
        # for count in counts:
        #     if(count != 0):
        #         return False

        s_dict = {}
        t_dict = {}

        for i in range(len(s)):
            s_dict[s[i]] = 1 + s_dict.get(s[i], 0)
            t_dict[t[i]] = 1 + t_dict.get(t[i], 0)
        
        return s_dict == t_dict